### PR TITLE
Feature/rework template mapping

### LINKED ISSUE #228: Hashing for Class-Template-Mapping descissions
Create a hash of the classes or a resource (sorted list of IRIs) and map it to the template decisions.
Produce warnings regarding the template decision only once.

### LINKED ISSUE #231: Output warning about multiple templates only if it is more then one.
The class mapper should only create a warning if the possible templates are actually more then one.

### LINKED ISSUE #244: Template selection precedence
If we have the following class template mappings:

```yaml
class_template_mappings:
  "http://xmlns.com/foaf/0.1/Person": "person"
  "http://example.org/Student": "student"
  "http://example.org/Professor": "professor"
```

this data
```turtle
@prefix ex: <http://example.org/> .
@prefix foaf: <http://xmlns.com/foaf/0.1/> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .

ex:Student rdfs:subClassOf foaf:Person .
ex:Professor rdfs:subClassOf foaf:Person .

ex:personA a foaf:Person .
ex:personB a foaf:Person, ex:Student .
ex:personC a foaf:Person, ex:Professor .
ex:personD a foaf:Person, ex:Student, ex:Professor .
```

than `http://example.org/personA` should be rendered with `person`, `http://example.org/personB` should be rendered with `student`, `http://example.org/personC` should be rendered with `professor`, and `http://example.org/personD` should produce a warning and should be rendered randomly with `professor` or `student`.

### PR DESCRIPTION
**Pull request by Simaris**
- [x] Hashing
- [x] local Search on template request
- [x] restore old behavior
  - [x] immediate template inheritance
  - [x] tests
  - [x] hookup warning system
- [x] add comments

Fix #231, fix #228, fix #244.

### CODE REVIEW

File: lib/jekyll/helper/rdf_class_extraction.rb [hunk 1]
```diff
@@ -2,47 +2,158 @@ module Jekyll
...
+            end
 
-          @classResources.each{|key, value|
-            value.find_direct_subclasses.each{|s|
-              value.add_subclass(@classResources[s])
+            class_resource.find_direct_superclasses.each{ |uri|
+              @classResources[uri] ||= Jekyll::JekyllRdf::Drops::RdfResourceClass.new(RDF::URI(uri))
+              if(!@classResources[uri].template.nil?) # do not search in paths you previously found
+                if @classResources[uri].base
+                  if(min_class.nil?)
```
Comment (Simaris) [inline lib/jekyll/helper/rdf_class_extraction.rb]: this is missing a lock < min_class_lock

**Comment by white-gecko** [comment]
classes_edit svg

Please test all cases which are depicted in the figure.

**Comment by white-gecko** [comment]
Is it possible to test if the `penguin` case produces a warning?

**Comment by white-gecko** [comment]
You should also make sure, that no warning is produced for the `fish`.

**Comment by Simaris** [comment]
I have corrected the fish test.

**Comment by white-gecko** [comment]
Is this issue now completed?

**Pull request merged**